### PR TITLE
Download dependencies when the `downloadDeps` option is passed to the vite config

### PR DESCRIPTION
**Pull request by JayaKrishnaNamburu**
The goal is to provide an option to the users to download the dependencies when running a build. Users can pass a flag called `downloadDeps` with which all the url's pointing to CDN's are downloaded.
And in `develoment` mode still serve the dependencies from the CDN for speed. 

- Serve deps from CDN in development mode.
- Being able to have a flag to download the deps when bundling.

Things to do

- [x] The dev-server is broken now for the deps from scope
- [x] Serve deps from CDN in development mode.

### CODE REVIEW

**Comment by JayaKrishnaNamburu** [comment]
Please don't merge it, i am adding tests. And found when the `inputMap` is not passed and the `downloadDeps` is specified. The plugins is breaking !!

**Comment by Aslemammad** [comment]
```
diff --git a/plugin/src/index.ts b/plugin/src/index.ts
index 4514a67..ceb8708 100644
--- a/plugin/src/index.ts
+++ b/plugin/src/index.ts
@@ -1,10 +1,16 @@
 import path from "path";
-import { Generator, GeneratorOptions } from "@jspm/generator";
-import type { ConfigEnv, Plugin, ResolvedConfig } from "vite";
+import { Generator, GeneratorOptions, fetch } from "@jspm/generator";
+import type {
+  ConfigEnv,
+  HtmlTagDescriptor,
+  Plugin,
+  ResolvedConfig,
+} from "vite";
 
 type PluginOptions = GeneratorOptions & {
   development?: boolean;
   strictInputMap?: boolean;
+  downloadDeps?: boolean;
 };
 
 const getDefaultOptions = (env: ConfigEnv): PluginOptions => ({
@@ -37,7 +43,6 @@ function getGenerator(options: PluginOptions) {
 }
 
 function plugin(_options?: PluginOptions): Plugin[] {
-  const installPromiseCache: Promise<unknown>[] = [];
   let resolvedConfig: ResolvedConfig;
   const resolvedDeps: Set<string> = new Set();
   let env: ConfigEnv;
@@ -70,11 +75,21 @@ function plugin(_options?: PluginOptions): Plugin[] {
           },
         } as Plugin);
       },
-      async resolveId(id, _, ctx) {
+      async resolveId(id, importer, ctx) {
+        const options = getOptions(env, _options);
         if (ctx.ssr) {
           // No plans on getting this working in SSR
-          return;
+          return null;
+        }
+
+        if (importer?.startsWith("http") && id?.startsWith(".")) {
+          const newPath = new URL(id, importer).toString();
+          if (options?.downloadDeps) {
+            return { id: newPath, external: false };
+          }
+          return { id, external: true };
         }
+
         if (
           id.startsWith("/") ||
           id.startsWith(".") ||
@@ -85,43 +100,96 @@ function plugin(_options?: PluginOptions): Plugin[] {
         ) {
           return null;
         }
-        const options = getOptions(env, _options);
+
+        const generator = getGenerator(options);
+
+        // if the module is resolved, ignore it, for cases like when inputMap
+        // option of jspm is used
+        let resolvedInInputMap = false;
+        let proxyImport;
+
         // if true and inputMap is defined in jspm options, we skip installing deps
         if (options.strictInputMap && options.inputMap) {
+          if (options?.downloadDeps) {
+            try {
+              proxyImport = generator.resolve(id);
+              resolvedDeps.add(id);
+              return { id: proxyImport, external: false };
+            } catch {
+              if (importer?.startsWith("http")) {
+                proxyImport = generator.importMap.resolve(id, importer);
+                return { id: proxyImport, external: false };
+              }
+            }
+          }
+
           return {
             id,
             external: true,
           };
         }
 
-        const generator = getGenerator(options);
-
-        // if the module is resolved, ignore it, for cases like when inputMap
-        // option of jspm is used
-        let resolvedInInputMap = false;
         try {
-          generator.resolve(id);
+          proxyImport = generator.resolve(id);
           resolvedInInputMap = true;
           resolvedDeps.add(id);
-        } catch {}
+        } catch {
+          if (importer?.startsWith("http")) {
+            proxyImport = generator.importMap.resolve(id, importer);
+            if (options?.downloadDeps) {
+              return { id: proxyImport, external: false };
+            }
+          }
+        }
 
         if (options?.development && env.command === "serve") {
           if (!resolvedInInputMap) {
             await generator.install(id);
+            proxyImport = generator.resolve(id);
           }
           resolvedDeps.add(id);
 
+          if (options?.downloadDeps) {
+            return {
+              id: proxyImport,
+              external: false,
+            };
+          }
+
           return {
             id,
             external: true,
           };
         }
+
+        console.log('here 1')
         if (!resolvedInInputMap) {
-          installPromiseCache.push(generator.install(id));
+          console.log('here2', id)
+          await generator.install(id);
+          console.log('here3', id, generator.getMap())
+          proxyImport = generator.resolve(id);
+          console.log('here4', proxyImport)
+        }
+        console.log('here',id, proxyImport)
+
+        if (options?.downloadDeps && process.env?.NODE_ENV === "production") {
+          return { id: proxyImport, external: false };
         }
 
         return { id, external: true };
       },
+      async load(id) {
+        const options = getOptions(env, _options);
+
+        if (!options?.downloadDeps) {
+          return;
+        }
+
+        if (id?.startsWith("http")) {
+          const code = await (await fetch(id)).text();
+          return code;
+        }
+      },
     },
     {
       name: "jspm:post",
@@ -132,31 +200,38 @@ function plugin(_options?: PluginOptions): Plugin[] {
         async transform(html) {
           const options = getOptions(env, _options);
           const generator = getGenerator(options);
-          await Promise.all(installPromiseCache);
-          resolvedDeps.clear()
-          installPromiseCache.length = 0;
+          resolvedDeps.clear();
+
+          const tags: HtmlTagDescriptor[] = [
+            {
+              tag: "script",
+              attrs: {
+                type: "module",
+                src: "https://ga.jspm.io/npm:es-module-shims@1.5.9/dist/es-module-shims.js",
+                async: !(options.development && env.command === "serve"),
+              },
+              injectTo: "head-prepend",
+            },
+          ];
+
+          if (
+            // only when we are in development or non-downloadDeps (prod-dev)
+            !options?.downloadDeps &&
+            process.env?.NODE_ENV !== "production"
+          ) {
+            tags.unshift({
+              tag: "script",
+              attrs: {
+                type: "importmap",
+              },
+              children: JSON.stringify(generator.getMap(), null, 2),
+              injectTo: "head-prepend",
+            });
+          }
 
           return {
             html,
-            tags: [
-              {
-                tag: "script",
-                attrs: {
-                  type: "importmap",
-                },
-                children: JSON.stringify(generator.getMap(), null, 2),
-                injectTo: "head-prepend",
-              },
-              {
-                tag: "script",
-                attrs: {
-                  type: "module",
-                  src: "https://ga.jspm.io/npm:es-module-shims@1.5.9/dist/es-module-shims.js",
-                  async: !(options.development && env.command === "serve"),
-                },
-                injectTo: "head-prepend",
-              },
-            ],
+            tags,
           };
         },
       },

```
Could you take a look at this diff, please!

File: plugin/src/index.ts [hunk 1]
```diff
@@ -70,11 +75,21 @@ function plugin(_options?: PluginOptions): Plugin[] {
           },
         } as Plugin);
       },
-      async resolveId(id, _, ctx) {
+      async resolveId(id, importer, ctx) {
+        const options = getOptions(env, _options);
         if (ctx.ssr) {
           // No plans on getting this working in SSR
           return;
```
Comment (Aslemammad) [inline plugin/src/index.ts]: It'd be better to return null here.

Comment (Aslemammad) [inline plugin/src/index.ts]: It'd be better to return null here.

File: plugin/src/index.ts [hunk 2]
```diff
@@ -132,31 +194,37 @@ function plugin(_options?: PluginOptions): Plugin[] {
...
+                src: "https://ga.jspm.io/npm:es-module-shims@1.5.9/dist/es-module-shims.js",
+                async: !(options.development && env.command === "serve"),
+              },
+              injectTo: "head-prepend",
+            },
+          ];
+
+          if (
+            !options?.downloadDeps ||
+            process.env?.NODE_ENV !== "production"
```
Comment (Aslemammad) [inline plugin/src/index.ts:198]:
```ts
          if (
            // only when we are in development or non-downloadDeps (prod-dev)
            !options?.downloadDeps &&
            process.env?.NODE_ENV !== "production"
          ) {

```
This is better.

**Review by Aslemammad — CHANGES_REQUESTED** [review]

**Comment by JayaKrishnaNamburu** [comment]
@Aslemammad i have the issue with the `dev` and `downloadDeps` yesterday. All good now, the removed the `strictImportMap` flag now. Since we are passing `inputMap` and after the next release of JSPM https://github.com/jspm/generator/pull/144

The `inputMap` is handled as source of truth in re-constructing the map. So, we don't need to enforce externally. The deps will be resolved only from the map that users provided.

And regarding the tests, all the tests passes in local machine. But the server and port from vite-test are quite random. Some test case fails at some odd time. But, all of them are good 👍

**Pull request merged**